### PR TITLE
Migrating fire over to the new entity system and improving the experience

### PR DESCRIPTION
**Pull request by Sunrunner37**
Currently fires in Nitrox are entirely hit and miss.  Players will douse fires in the aurora but the behavior doesn't sync correctly due to prefab spawning issues.  Similarly, fires in the cyclops will spawn about 50% of the time and have no ability to persist after a server reset or reconnect.  The smoke would also hang around for some players.  

This PR migrates fires to the new entity system and deprecates 8 fire-specific classes.

Fires put out after reconnection:

image

Cooperative fire fighting:

image

Fires in the cyclops:

image

Smoke clearing after the fire: 

image

### CODE REVIEW

**Review by dartasen — APPROVED** [review]

File: NitroxPatcher/Patches/Dynamic/Fire_Douse_Patch.cs [hunk 1]
```diff
@@ -1,30 +1,73 @@
...
-            if (!__instance.livemixin.IsAlive() || __instance.IsExtinguished())
-            {
-                NitroxServiceLocator.LocateService<Fires>().OnDouse(__instance, 10000);
-            }
-            else
-            {
-                NitroxServiceLocator.LocateService<Fires>().OnDouse(__instance, amount);
-            }
+#if DEBUG
+            Log.Error($"Fire instance is unknown to the server {__instance.gameObject.name}");
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/Fire_Douse_Patch.cs:24]: Server?

File: NitroxPatcher/Patches/Dynamic/Fire_Douse_Patch.cs [hunk 2]
```diff
@@ -1,30 +1,73 @@
...
-        public override void Patch(Harmony harmony)
+        if (!__instance.livemixin.IsAlive() || __instance.IsExtinguished())
+        {
+            // Ensure one final metadata packet goes out to trigger any residual extinguish event that need to happen.
+            // We want to make sure this packet happens before the delete, so send it directly and not with throttling.
+            Resolve<ThrottledPacketSender>().RemovePendingPackets(nitroxId);
+            Resolve<Entities>().EntityMetadataChanged(__instance, nitroxId);
+
+            // Then tell the serve to delete the entity as it is now extinguished. 
+            Resolve<IPacketSender>().Send(new EntityDestroyed(nitroxId));
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/Fire_Douse_Patch.cs:37]: Is there a chance for race conditions and could it cause problems? Maybe we want to wait a little before deleting the entity?

File: NitroxPatcher/Patches/Dynamic/SubFire_CreateFire_Patch.cs [hunk 3]
```diff
@@ -34,14 +39,19 @@ public static void Postfix(SubFire __instance, SubFire.RoomFire startInRoom, boo
...
                     if (transform.childCount > 0)
                     {
                         int nodeIndex = Array.IndexOf(__instance.roomFires[startInRoom.roomLinks.room].spawnNodes, transform);
                         Fire fire = transform.GetComponentInChildren<Fire>();
-                        fires.OnCreate(fire, startInRoom, nodeIndex);
+
+                        NitroxId subRootId = NitroxEntity.GetId(fire.fireSubRoot.gameObject);
+                        NitroxId fireId = NitroxEntity.GetId(fire.gameObject);
+
+                        CyclopsFireEntity cyclopsFireEntity = new((int)startInRoom.roomLinks.room, nodeIndex, fireId, TechType.None.ToDto(), null, subRootId, new());
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/SubFire_CreateFire_Patch.cs:52]:
```suggestion
                        CyclopsFireEntity cyclopsFireEntity = new((int)startInRoom.roomLinks.room, nodeIndex, fireId, NitroxTechType.None, null, subRootId, new());
```

**Review by Jannify — CHANGES_REQUESTED** [review]
`WorldPersistenceTest` is missing checks for `CyclopsFireEntity` and `FireMetadata`.

File: NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    private IEnumerator SpawnFire(SubFire subFire, CyclopsRooms room, int nodeIndex, NitroxId id, TaskResult<Optional<GameObject>> result)
+    {
+        SubFire.RoomFire roomFire = subFire.roomFires[room];
+        Transform fireTransform = roomFire.spawnNodes[nodeIndex];
+
+        // If a fire already exists at the node, replace the old Id with the new one
+        if (fireTransform.childCount > 0)
+        {
+            Fire existingFire = fireTransform.GetComponentInChildren<Fire>();
```
Comment (tornac1234) [inline NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs:45]:
```suggestion
            Fire existingFire = fireTransform.GetComponentInChildren<Fire>(true);
```

File: NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+            if (existingFire && NitroxEntity.GetId(existingFire.gameObject) != id)
```
Comment (tornac1234) [inline NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs:47]: Change ``GetId`` to something more relevant (after rebase)

File: NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        GameObject newFireGameObject = null;
+
+        firePrefab.SpawnManual(delegate (GameObject fireGO)
+        {
+            newFireGameObject = fireGO;
+        });
+
+        yield return new WaitUntil(() => newFireGameObject);
+
+        Fire fire = newFireGameObject.GetComponentInChildren<Fire>();
```
Comment (tornac1234) [inline NitroxClient/GameLogic/Spawning/CyclopsFireEntitySpawner.cs:70]:
```suggestion
        Fire fire = newFireGameObject.GetComponentInChildren<Fire>(true);
```

**Review by tornac1234 — CHANGES_REQUESTED** [review]
Needs rebase, modifications from #2063 (patches) and #1908 (replacing Nitrox.GetId) should be applied.

**Comment by dartasen** [comment]
Will need few rework and rebase to be properly integrated, closing it for now

**Pull request closed without merging**